File: js/AgentsOrchestrator/Repo.js
import { git } from '../gitHelpers/git.js';
import { Branch } from './Branch.js';
import { Logger, resolveLevel } from '../utils/logger.js';
import { parseCommitMessage } from '../gitHelpers/parseCommitMessage.js';

/**
 * Repo class to manage Git repositories.
 *
 * It:
 * 1. Initializes with a configuration object.
 * 2. Fetches branches (local by default, or remote if aynigRemote is set).
 * 3. Creates Branch instances for each branch.
 * 4. Runs each Branch instance.
 * 5. waits for all Branch instances to complete.
 *
 * It has no clue about what each Branch does internally.
 */
export class Repo {
    constructor(config) {
        this.config = config;
    }

    filterBranches(branches, current, mode) {
        switch (mode) {
            case 'skip':
                return current ? branches.filter(name => name !== current) : branches;
            case 'only':
                return current ? [current] : [];
            default:
                return branches;
        }
    }

    async resolveCurrentRemoteBranch(localCurrent) {
        if (!localCurrent) {
            return '';
        }
        try {
            const upstream = (await git.raw(['rev-parse', '--abbrev-ref', `${localCurrent}@{upstream}`])).trim();
            if (!upstream) {
                return '';
            }
            if (!upstream.startsWith(`${this.config.aynigRemote}/`)) {
                this.config.logger.warn('Current branch upstream %s does not belong to remote %s', upstream, this.config.aynigRemote);
                return '';
            }
            return upstream;
        } catch {
            return '';
        }
    }

    remoteFromTrailers(trailers = {}) {
        const raw = trailers['aynig-remote'];
        if (Array.isArray(raw)) {
            return String(raw[0] || '').trim();
        }
        return String(raw || '').trim();
    }

    async resolveAynigRemoteFromHead() {
        if (this.config.aynigRemote) {
            return this.config.aynigRemote;
        }

        try {
            const raw = await git.raw(['log', '-1', '--pretty=format:%s%x1f%b']);
            const parts = raw.split('\x1f');
            const message = (parts[0] || '').replace(/\n+$/, '');
            const body = (parts[1] || '').replace(/\n+$/, '');
            const parsed = await parseCommitMessage({ message, body });
            return this.remoteFromTrailers(parsed.trailers);
        } catch {
            return '';
        }
    }

    async run() {
        if (!this.config.logger) {
            const baseLevel = resolveLevel({
                cliLevel: this.config.logLevel,
                cliSet: this.config.logLevelSet,
                trailerLevel: '',
                envLevel: process.env.AYNIG_LOG_LEVEL,
                defaultLevel: this.config.logLevel
            });
            this.config.logger = new Logger(baseLevel);
        }

        // Resolve repo root once and share it via config
        this.config.repoRoot = await git.revparse(['--show-toplevel']);
        this.config.logger.info('Repository root: %s', this.config.repoRoot.trim());

        this.config.aynigRemote = await this.resolveAynigRemoteFromHead();
        if (this.config.aynigRemote) {
            this.config.logger.info('Fetching remote branches from %s', this.config.aynigRemote);
            await git.fetch();
        }

        // Get remote branches if aynigRemote is set, otherwise get local branches
        const branchesInfo = this.config.aynigRemote
            ? await git.branch(['-r'])
            : await git.branchLocal();

        const localInfo = await git.branchLocal();
        const localCurrent = localInfo.current || '';

        const filterCurrent = this.config.aynigRemote
            ? await this.resolveCurrentRemoteBranch(localCurrent)
            : localCurrent;

        if (this.config.aynigRemote && this.config.currentBranch === 'only' && !filterCurrent) {
            this.config.logger.warn('Current branch has no upstream for --current-branch=only in remote mode');
        }

        const branchNames = this.filterBranches(branchesInfo.all, filterCurrent, this.config.currentBranch || 'skip');
        this.config.logger.info('Running %d branches (current-branch=%s)', branchNames.length, this.config.currentBranch || 'skip');

        this.branches = branchNames.map(name => new Branch({
            config: this.config,
            branchName: name,
            isCurrentBranch: !this.config.aynigRemote && name === localCurrent,
        }));

        await Promise.all(this.branches.map(branch => branch.run()));
    }
}

File: go/internal/orchestrator/repo.go
package orchestrator

import (
	"os"
	"strings"
	"sync"

	"all-you-need-is-git/go/internal/config"
	"all-you-need-is-git/go/internal/gitx"
	"all-you-need-is-git/go/internal/logx"
)

type Repo struct {
	config config.Config
	logger logx.Logger
}

func NewRepo(cfg config.Config) *Repo {
	resolved := logx.ResolveLevel(cfg.LogLevel, cfg.LogLevelSet, "", os.Getenv("AYNIG_LOG_LEVEL"), config.Default().LogLevel)
	return &Repo{config: cfg, logger: logx.New(resolved)}
}

func (r *Repo) Run() error {
	repoRoot, err := gitx.RepoRoot()
	if err != nil {
		return err
	}
	r.config.RepoRoot = repoRoot
	r.logger.Infof("Repository root: %s", repoRoot)

	if r.config.UseRemote == "" {
		commit, readErr := gitx.ReadCommit("HEAD")
		if readErr == nil {
			if remote := firstLowerTrailerValue(commit.Trailers, "aynig-remote"); remote != "" {
				r.config.UseRemote = remote
				r.logger.Infof("Using remote from trailer aynig-remote=%s", remote)
			}
		}
	}

	if r.config.UseRemote != "" {
		r.logger.Infof("Fetching remote branches from %s", r.config.UseRemote)
		if err := gitx.Fetch(repoRoot); err != nil {
			return err
		}
	}

	current, err := gitx.BranchCurrent(repoRoot)
	if err != nil {
		return err
	}
	r.logger.Debugf("Current branch: %s", current)

	var branches []string
	if r.config.UseRemote != "" {
		branches, err = gitx.BranchListRemote(repoRoot)
	} else {
		branches, err = gitx.BranchListLocal(repoRoot)
	}
	if err != nil {
		return err
	}
	r.logger.Debugf("Found %d branches", len(branches))

	branchNames := filterBranches(branches, current, r.config.CurrentBranch)
	if r.config.UseRemote != "" {
		upstream, err := gitx.BranchUpstream(repoRoot, current)
		if err != nil {
			return err
		}
		if upstream != "" {
			if strings.HasPrefix(upstream, r.config.UseRemote+"/") {
				branchNames = filterBranches(branches, upstream, r.config.CurrentBranch)
			} else {
				r.logger.Warnf("Current branch upstream %s does not belong to remote %s", upstream, r.config.UseRemote)
				branchNames = filterBranches(branches, "", r.config.CurrentBranch)
			}
		} else {
			if r.config.CurrentBranch == "only" {
				r.logger.Warnf("Current branch has no upstream for --current-branch=only in remote mode")
			}
			branchNames = filterBranches(branches, "", r.config.CurrentBranch)
		}
	}
	r.logger.Infof("Running %d branches (current-branch=%s)", len(branchNames), r.config.CurrentBranch)

	var wg sync.WaitGroup
	errCh := make(chan error, len(branchNames))
	for _, name := range branchNames {
		branch := NewBranch(r.config, name, name == current)
		wg.Add(1)
		go func(b *Branch) {
			defer wg.Done()
			if err := b.Run(); err != nil {
				errCh <- err
			}
		}(branch)
	}
	wg.Wait()
	close(errCh)

	for err := range errCh {
		if err != nil {
			return err
		}
	}
	return nil
}

func filterBranches(all []string, current string, mode string) []string {
	switch mode {
	case "skip":
		out := []string{}
		for _, name := range all {
			if name != current {
				out = append(out, name)
			}
		}
		return out
	case "only":
		if current == "" {
			return []string{}
		}
		return []string{current}
	default:
		return all
	}
}

func firstLowerTrailerValue(trailers map[string][]string, key string) string {
	want := strings.ToLower(strings.TrimSpace(key))
	if want == "" {
		return ""
	}
	for k, values := range trailers {
		if strings.ToLower(strings.TrimSpace(k)) != want {
			continue
		}
		if len(values) == 0 {
			return ""
		}
		return strings.TrimSpace(values[0])
	}
	return ""
}
